File: scripts/red_team_status.py
"""CLI for displaying pending red-team human-review halts (FU-rt8).

Lists every human-review finding the operator has not yet accepted via
``red_team_accept`` so a single command answers "what is blocking me?".

Usage:

  python3 red_team_status.py                       # everything pending
  python3 red_team_status.py --repo Evinced/foo    # one repo
  python3 red_team_status.py --stage design        # one stage
  python3 red_team_status.py --json                # machine-readable

The display is read-only — it never mutates state. Acceptance happens via
``red_team_accept.py``.
"""

from __future__ import annotations

import argparse
import json
import sys
from dataclasses import asdict
from typing import Sequence

import red_team_human_review


def _format_human(halts: list, out=sys.stdout) -> None:
    if not halts:
        print("red_team_status: no pending human-review halts", file=out)
        return
    print(
        f"red_team_status: {len(halts)} pending human-review halts",
        file=out,
    )
    for h in halts:
        excerpt = (h.concern_excerpt or "").replace("\n", " ").strip()
        if len(excerpt) > 140:
            excerpt = excerpt[:137] + "..."
        repo_segment = f"  repo={h.repo}" if h.repo else ""
        pr_segment = f"  pr={h.pr_number}" if h.pr_number else ""
        path_segment = (
            f"  artifact={h.artifact_relative_path}"
            if h.artifact_relative_path
            else ""
        )
        print("", file=out)
        print(f"- stable_key: {h.stable_key}", file=out)
        print(
            f"  run_id={h.run_id}  stage={h.stage}  round={h.round_num}"
            f"  persona={h.persona}  finding_id={h.finding_id}"
            f"{repo_segment}{pr_segment}{path_segment}",
            file=out,
        )
        if excerpt:
            print(f"  concern: {excerpt}", file=out)


def _format_json(halts: list, out=sys.stdout) -> None:
    json.dump([asdict(h) for h in halts], out, indent=2, sort_keys=True)
    out.write("\n")


def _build_parser() -> argparse.ArgumentParser:
    p = argparse.ArgumentParser(
        prog="red_team_status",
        description=(
            "List pending red-team human-review halts. Use red_team_accept "
            "to acknowledge them by stable_key."
        ),
    )
    p.add_argument("--repo", default=None, help="Filter to one repo (nameWithOwner).")
    p.add_argument("--stage", default=None, choices=("design", "plan"), help="Filter by stage.")
    p.add_argument("--json", action="store_true", help="Emit JSON instead of human text.")
    p.add_argument(
        "--db",
        default=None,
        help=(
            "Audit DB path override. Defaults to the canonical resolver "
            "(scripts/red_team_audit_cli.py resolve-db)."
        ),
    )
    return p


def main(argv: Sequence[str] | None = None) -> int:
    args = _build_parser().parse_args(argv)
    from red_team_audit import resolve_db_path
    db_path = resolve_db_path(args.db)
    halts = red_team_human_review.list_pending_halts(
        repo=args.repo, stage=args.stage, db_path=db_path,
    )
    if args.json:
        _format_json(halts)
    else:
        _format_human(halts)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/red_team_status.ts
#!/usr/bin/env node
/**
 * `/red_team status` TS CLI — port of `scripts/red_team_status.py`.
 *
 * Lists every pending red-team human-review halt (any unaccepted
 * finding whose counter_proposal is REQUEST_HUMAN_REVIEW). Read-only.
 *
 * Usage:
 *   node --experimental-strip-types tools/red_team_status.ts
 *   node --experimental-strip-types tools/red_team_status.ts --repo Evinced/foo
 *   node --experimental-strip-types tools/red_team_status.ts --stage spec
 *   node --experimental-strip-types tools/red_team_status.ts --json
 *
 * Acceptance happens via `tools/red_team_accept.ts`.
 */

import path from "node:path";
import { fileURLToPath } from "node:url";

import { resolveDb } from "./red_team_db_resolver.ts";

import {
  type PendingHalt,
  listPendingHalts,
} from "./red_team_human_review_lib.ts";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const REPO_ROOT = path.resolve(HERE, "..");

interface CliArgs {
  repo: string | null;
  stage: "spec" | "plan" | null;
  json: boolean;
  db: string | null;
  help: boolean;
}

function parseArgs(argv: string[]): CliArgs {
  const args: CliArgs = {
    repo: null,
    stage: null,
    json: false,
    db: null,
    help: false,
  };
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i]!;
    const next = (): string => {
      const v = argv[++i];
      if (v === undefined) throw new Error(`${arg} requires a value`);
      return v;
    };
    switch (arg) {
      case "--repo":
        args.repo = next();
        break;
      case "--stage": {
        const v = next();
        if (v !== "spec" && v !== "plan") {
          throw new Error(`--stage must be spec|plan, got ${JSON.stringify(v)}`);
        }
        args.stage = v;
        break;
      }
      case "--json":
        args.json = true;
        break;
      case "--db":
        args.db = next();
        break;
      case "-h":
      case "--help":
        args.help = true;
        break;
      default:
        throw new Error(`unknown argument: ${arg}`);
    }
  }
  return args;
}

function printHelp(): void {
  process.stdout.write(
    `\
Usage: red_team_status [options]

List pending red-team human-review halts. Use red_team_accept to
acknowledge them by stable_key.

Options:
  --repo NAME        Filter to one repo (nameWithOwner).
  --stage STAGE      Filter by stage (spec|plan).
  --json             Emit JSON instead of human text.
  --db PATH          Audit DB path override. Defaults to the canonical
                     resolver (scripts/red_team_audit_cli.py resolve-db).
  -h, --help         Show this help.
`,
  );
}


function formatHuman(halts: readonly PendingHalt[]): string {
  if (halts.length === 0) {
    return "red_team_status: no pending human-review halts\n";
  }
  const lines: string[] = [
    `red_team_status: ${halts.length} pending human-review halts`,
  ];
  for (const h of halts) {
    let excerpt = (h.concern_excerpt ?? "").replace(/\n/g, " ").trim();
    if (excerpt.length > 140) excerpt = excerpt.slice(0, 137) + "...";
    const repoSeg = h.repo ? `  repo=${h.repo}` : "";
    const prSeg = h.pr_number ? `  pr=${h.pr_number}` : "";
    const pathSeg = h.artifact_relative_path
      ? `  artifact=${h.artifact_relative_path}`
      : "";
    lines.push("");
    lines.push(`- stable_key: ${h.stable_key}`);
    lines.push(
      `  run_id=${h.run_id}  stage=${h.stage}  round=${h.round_num}` +
        `  persona=${h.persona}  finding_id=${h.finding_id}${repoSeg}${prSeg}${pathSeg}`,
    );
    if (excerpt) lines.push(`  concern: ${excerpt}`);
  }
  return lines.join("\n") + "\n";
}

/** JSON serializer that matches Python `json.dump(..., indent=2, sort_keys=True)`. */
function formatJson(halts: readonly PendingHalt[]): string {
  return (
    JSON.stringify(
      halts.map((h) => sortKeys(h)),
      null,
      2,
    ) + "\n"
  );
}

function sortKeys<T>(value: T): T {
  if (Array.isArray(value)) return value.map(sortKeys) as T;
  if (value !== null && typeof value === "object") {
    const obj = value as Record<string, unknown>;
    const sorted: Record<string, unknown> = {};
    for (const k of Object.keys(obj).sort()) sorted[k] = sortKeys(obj[k]);
    return sorted as T;
  }
  return value;
}

export function main(argv: string[]): number {
  let args: CliArgs;
  try {
    args = parseArgs(argv);
  } catch (err) {
    process.stderr.write(`red_team_status: ${(err as Error).message}\n`);
    return 2;
  }
  if (args.help) {
    printHelp();
    return 0;
  }
  const dbPath = resolveDb(args.db).db_path;
  const halts = listPendingHalts({
    repo: args.repo,
    stage: args.stage,
    dbPath,
  });
  if (args.json) {
    process.stdout.write(formatJson(halts));
  } else {
    process.stdout.write(formatHuman(halts));
  }
  return 0;
}

if (import.meta.url === `file://${process.argv[1]}`) {
  process.exit(main(process.argv.slice(2)));
}
